typescript/agentkit/src/action-providers/spl/splActionProvider.ts

```diff
@@ -0,0 +1,143 @@
+  /**
+   * Creates a new SplActionProvider instance.
+   */
+  constructor() {
+    super("spl", []);
+  }
+
+  /**
+   * Transfer SPL tokens to another address.
+   *
+   * @param walletProvider - The wallet provider to use for the transfer
+   * @param args - Transfer parameters including recipient address, mint address, and amount
+   * @returns A message indicating success or failure with transaction details
+   */
+  @CreateAction({
+    name: "transfer",
+    description: `
+    This tool will transfer SPL tokens to another address.
+    - Amount should be specified in token units (not raw)
+    - Recipient must be a valid Solana address
+    - Mint address must be a valid SPL token mint
+    - Ensures sufficient token balance before transfer
+    - Returns transaction details
+    `,
+    schema: TransferTokenSchema,
+  })
+  async transfer(
+    walletProvider: SvmWalletProvider,
+    args: z.infer<typeof TransferTokenSchema>,
+  ): Promise<string> {
+    try {
+      const connection = walletProvider.getConnection();
+      const fromPubkey = walletProvider.getPublicKey();
+      const toPubkey = new PublicKey(args.recipient);
+      const mintPubkey = new PublicKey(args.mintAddress);
+
+      const mintInfo = await getMint(connection, mintPubkey);
+      const adjustedAmount = args.amount * Math.pow(10, mintInfo.decimals);
+
+      const sourceAta = await getAssociatedTokenAddress(mintPubkey, fromPubkey);
+      const destinationAta = await getAssociatedTokenAddress(mintPubkey, toPubkey);
+
+      const instructions: TransactionInstruction[] = [];
+
+      try {
+        await getAccount(connection, sourceAta);
+      } catch {
+        throw new Error(
+          "Source token account does not exist. You need tokens before you can transfer them.",
+        );
+      }
+
+      try {
+        await getAccount(connection, destinationAta);
+      } catch {
+        instructions.push(
+          createAssociatedTokenAccountInstruction(fromPubkey, destinationAta, toPubkey, mintPubkey),
+        );
+      }
+
+      const sourceAccount = await getAccount(connection, sourceAta);
+      if (sourceAccount.amount < BigInt(adjustedAmount)) {
+        throw new Error(
+          `Insufficient token balance. Have ${sourceAccount.amount}, need ${adjustedAmount}`,
+        );
+      }
+
+      instructions.push(
+        createTransferCheckedInstruction(
+          sourceAta,
+          mintPubkey,
+          destinationAta,
+          fromPubkey,
+          adjustedAmount,
+          mintInfo.decimals,
+        ),
+      );
+
+      const tx = new VersionedTransaction(
+        MessageV0.compile({
+          payerKey: fromPubkey,
+          instructions: instructions,
+          recentBlockhash: (await connection.getLatestBlockhash()).blockhash,
+        }),
+      );
+
+      const signedTx = await walletProvider.signTransaction(tx);
+      const txHash = await walletProvider.sendTransaction(signedTx);
```

Comment:
I think the correct approach here is to use `walletProvider.signAndSendTransaction` which is canonical in Solana development:
* See Privy SDK docs: https://docs.privy.io/guide/server-wallets/usage/solana#using-the-sdk
* See Phantom docs: https://docs.privy.io/guide/server-wallets/usage/solana#using-the-sdk